hs_tools_resource/app_launch_helper.py

```diff
@@ -35,14 +35,35 @@ def resource_level_tool_urls(resource_obj, request_obj):
             continue
         if tool_metadata.supported_resource_types.supported_res_types.filter(
                 description__iexact=resource_obj.resource_type).exists():
-            if _check_user_can_view_app(request_obj, tool_res_obj) and \
-                    _check_app_supports_resource_sharing_status(resource_obj, tool_res_obj):
-
-                tl = _get_app_tool_info(request_obj, resource_obj, tool_res_obj)
-                if tl:
-                    tool_list.append(tl)
-                    if tl['url']:
-                        resource_level_app_counter += 1
+            if not _check_user_can_view_app(request_obj, tool_res_obj):
+                continue
+            if not _check_app_supports_resource_sharing_status(resource_obj, tool_res_obj):
+                continue
+
+            supported_file_extensions = _get_tools_supported_file_extensions(tool_res_obj)
+            supported_aggregation_types = _get_tools_supported_aggregation_types(tool_res_obj)
+            resource_contains_supported_file_extensions = _check_resource_contains_supported_file_extensions(
+                resource_obj, supported_file_extensions)
+            resource_contains_supported_aggregation_types = _check_resource_contains_supported_aggregation_types(
+                resource_obj, supported_aggregation_types)
+
+            # if the tool declares file extensions, then the resource must contain at least one of them
+            # unless the resource contains supported aggregation types
+            if supported_file_extensions and not resource_contains_supported_file_extensions:
+                if not supported_aggregation_types or not resource_contains_supported_aggregation_types:
+                    continue
+
+            # if the tool declares supported aggregation types, then the resource must contain at least one of them
+            # unless the resource contains supported file extensions
```

Comment:
Similarly, if a tool requires certain aggregations types and the resource doesn't have at least one of those aggregation types then that tool is not suitable for the resource.

If the tool has extension as well as aggregation type requirements, then the resource must meet both those requirements.